### PR TITLE
Immutable scim-schema objects

### PR DESCRIPTION
**Pull request by wallner**
This commit makes all scim-schema classes into immutable. This way the
handling in multi threaded environments gets way easier. Closes #167.

### CODE REVIEW

File: src/main/java/org/osiam/resources/scim/Address.java [hunk 1]
```diff
@@ -36,25 +37,49 @@
...
+        this.region = region;
+        this.postalCode = postalCode;
+        this.country = country;
+        this.type = type;
     }
 
+    /**
+     * Constructor for serialization. It is not intended for public use.
+     */
+
```
Comment (fwilhe) [inline src/main/java/org/osiam/resources/scim/Address.java]: Please remove the empty lines between the comment and the constructor itself.

File: src/main/java/org/osiam/resources/scim/Group.java [hunk 2]
```diff
@@ -47,18 +50,24 @@
...
+    @JsonCreator
+    private Group(@JsonProperty("id") String id,
+                  @JsonProperty("externalId") String externalId,
+                  @JsonProperty("meta") Meta meta,
+                  @JsonProperty(value = "schemas", required = true) Set<String> schemas,
+                  @JsonProperty("displayName") String displayName,
+                  @JsonProperty("members") Set<MemberRef> members) {
+        super(id, externalId, meta, schemas);
+        this.displayName = displayName;
+        this.members = (members != null ? members : new HashSet<MemberRef>());
```
Comment (fwilhe) [inline src/main/java/org/osiam/resources/scim/Group.java]: Is there a reason why you don't use [`Collections.emptySet()`](http://docs.oracle.com/javase/7/docs/api/java/util/Collections.html?is-external=true#emptySet()) here?

File: src/main/java/org/osiam/resources/scim/Entitlement.java [hunk 3]
```diff
@@ -22,30 +22,39 @@
...
+        this.type = type;
+    }
 
     /**
-     * Default constructor for Jackson
+     * Constructor for deserialization, it is not intended for general use.
      */
-    private Entitlement() {
-    }
+
```
Comment (fwilhe) [inline src/main/java/org/osiam/resources/scim/Entitlement.java]: Please remove the empty lines between the comment and the constructor itself.

**Comment by fwilhe** [comment]
According to the [javadoc](https://github.com/FasterXML/jackson-annotations/blob/93ae86c276ebb8327b2addc09d9e02035bf15d48/src/main/java/com/fasterxml/jackson/annotation/JsonProperty.java#L15), the default value of `@JsonProperty` is the name of the field. Is there a reason why you specify this exact name for the fields? I can see that it is a valid decision to declare this explicitly, but I wonder if there is any reasoning beyond just "making it explicit", rather then relying on the default behavior?

**Comment by fwilhe** [comment]
The class `Extension` is not declared `final`. Does it have to be subclassed?

File: src/main/java/org/osiam/resources/scim/Group.java (hunk 2 above)
Comment (wallner) [inline src/main/java/org/osiam/resources/scim/Group.java]: No. Good catch.

**Comment by wallner** [comment]
We want the scim-schema to be usable with at least java 7, before Java 8 it is not possible to access method parameter names via reflection, so to help Jackson along we need to provide the information in the annotation. As soon as we decide to drop java 7 support in the connector we can get rid of the name in the annotation.

**Comment by wallner** [comment]
There is no reason for `Extension` and `Role` to not be final. I fixed that. Thank you for spotting that.

**Comment by fwilhe** [comment]
Ok, this looks good to me now. Thanks a lot for your effort.

**Pull request merged**